src/circuit/gadget/sinsemilla/note_commit.rs

```diff
@@ -0,0 +1,1721 @@
+        // - 0 ≤ g_1 + (2^9)g_2 + 2^130 - t_P < 2^130 (13 ten-bit lookups)
+
+        // Decompose the low 130 bits of g1_g2_prime = g_1 + (2^9)g_2 + 2^130 - t_P,
+        // and output the running sum at the end of it.
+        // If g1_g2_prime < 2^130, the running sum will be 0.
+        let g1_g2_prime = g_1.value().zip(g_2.value()).map(|(g_1, g_2)| {
+            let two_pow_9 = pallas::Base::from_u64(1u64 << 9);
+            let two_pow_130 = pallas::Base::from_u128(1u128 << 65).square();
+            let t_p = pallas::Base::from_u128(T_P);
+            g_1 + (two_pow_9 * g_2) + two_pow_130 - t_p
+        });
+
+        let zs = self.sinsemilla_config.lookup_config.witness_check(
+            layouter.namespace(|| "Decompose low 130 bits of (g_1 + (2^9)g_2 + 2^130 - t_P)"),
+            g1_g2_prime,
+            13,
+            false,
+        )?;
+        let g1_g2_prime = zs[0];
+        assert_eq!(zs.len(), 14); // [z_0, z_1, ..., z_13]
+
+        Ok((g1_g2_prime, zs[13]))
+    }
+
+    // Check canonicity of y-coordinate given its LSB as a value.
+    // Also, witness the LSB and return the witnessed cell.
+    fn y_canonicity(
+        &self,
+        mut layouter: impl Layouter<pallas::Base>,
+        y: CellValue<pallas::Base>,
+        lsb: Option<pallas::Base>,
+    ) -> Result<CellValue<pallas::Base>, Error> {
+        // Decompose the field element
+        //      y = LSB || k_0 || k_1 || k_2 || k_3
+        //        = (bit 0) || (bits 1..=9) || (bits 10..=249) || (bits 250..=253) || (bit 254)
+        let (k_0, k_1, k_2, k_3) = {
+            let k_0 = y.value().map(|y| bitrange_subset(y, 1..10));
+            let k_1 = y.value().map(|y| bitrange_subset(y, 10..250));
+            let k_2 = y.value().map(|y| bitrange_subset(y, 250..254));
+            let k_3 = y.value().map(|y| bitrange_subset(y, 254..255));
+
+            (k_0, k_1, k_2, k_3)
+        };
+
+        // Range-constrain k_0 to be 9 bits.
+        let k_0 = self.sinsemilla_config.lookup_config.witness_short_check(
+            layouter.namespace(|| "Constrain k_0 to be 9 bits"),
+            k_0,
+            9,
+        )?;
+
+        // Range-constrain k_2 to be 4 bits.
+        let k_2 = self.sinsemilla_config.lookup_config.witness_short_check(
+            layouter.namespace(|| "Constrain k_2 to be 4 bits"),
+            k_2,
+            4,
+        )?;
+
+        // Decompose j = LSB + (2)k_0 + (2^10)k_1 using 25 ten-bit lookups.
+        let (j, z1_j, z13_j) = {
+            let j = lsb.zip(k_0.value()).zip(k_1).map(|((lsb, k_0), k_1)| {
+                let two = pallas::Base::from_u64(2);
+                let two_pow_10 = pallas::Base::from_u64(1 << 10);
+                lsb + two * k_0 + two_pow_10 * k_1
+            });
+            let zs = self.sinsemilla_config.lookup_config.witness_check(
+                layouter.namespace(|| "Decompose j = LSB + (2)k_0 + (2^10)k_1"),
+                j,
+                25,
+                false,
```

Comment:
I think this has to be a strict decomposition (since it is replacing the decomposition usually done by the SinsemillaHash).

```suggestion
                true,
```

Reply:
Actually I think we could just do a 13-word lookup decomposition here, and redistribute the pieces as 
```
(bit 0) || (bits 1..=9) || (bits 10..=129) || (bits 130..=253) || bit 254
```

Reply:
If bits 130..=253 are not range-constrained, how do you know that bits 0..=254 are actually a 10-bit-word decomposition of y(g<sub>d</sub>) or y(pk<sub>d</sub>)?